### PR TITLE
Feat: random channel dropout

### PR DESCRIPTION
**Pull request by vgilabert94**
#### Changes
Re-open a past augmentation problem: RandomChannelDropout. 
#2594 #2240 #2237 

related to #941 

This implementation not follow albumentation due to batching. 
For now num_drop_channels is fixed and controled by user (optional) for entire batch. The problem here is I don't know how to batch index using the code I have now. I am open to advice on how to improve this section.

Example:
image

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/augmentation/_2d/intensity/channel_dropout.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        # Generator of random parameters.
+        self._param_generator = ChannelDropoutGenerator(self.num_drop_channels)
+
+    def apply_transform(
+        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any], transform: Optional[Tensor] = None
+    ) -> Tensor:
+        KORNIA_CHECK_SHAPE(input, ["*", "3", "H", "W"])
+
+        out = input.clone()
+        out[params["dropout_mask"].to(device=out.device)] = self.fill_value
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/channel_dropout.py:87]: @vgilabert94 have you benchmark tried directly just to keep the indices (instead of allocating a mask) and just torch fill_ ? @johnnv1 this could be a good use case to start putting in practice the ops benchmark api we discussed

Comment (vgilabert94) [reply kornia/augmentation/_2d/intensity/channel_dropout.py:87]: If you can provide me a snippet code or some example of how i can do it.

Comment (johnnv1) [reply kornia/augmentation/_2d/intensity/channel_dropout.py:87]:
```python
fill_value = 0
inpt = torch.rand(1,4, 5, 3)
channels_to_fill = [0, 2]
out = inpt.clone()
return out[..., channels_to_fill, :, : ] = fill_value
```

Comment (johnnv1) [reply kornia/augmentation/_2d/intensity/channel_dropout.py:87]: about the performance, i remember seeing something in an old PR, but I didn't find it 😞 

you can benchmark your augmentation with https://pytorch.org/tutorials/recipes/recipes/benchmark.html -- here: https://github.com/kornia/kornia-benchmark  have some examples  

will look something like
```python
from kornia.augmentation import RandomChannelDropout
import torch
from torch.utils import benchmark


aug = RandomChannelDropout()
inpt = torch.rand(5, 3, 512, 512, device=torch.device('cuda'))

result = benchmark.Timer(
    stmt="aug(inpt)",
    setup="from __main__ import aug",
    globals={
        "image": inpt,
    },
    description='RandomChannelDropout',
).timeit(1000)

print(result)
```

if you have two cases/results to check you can do  

```python
from kornia.augmentation import RandomChannelDropout
import torch
from torch.utils import benchmark

aug_A = RandomChannelDropout_A()
aug_B = RandomChannelDropout_B()
inpt = torch.rand(5, 3, 512, 512, device=torch.device('cuda'))

results = [
  benchmark.Timer(
      stmt="aug_A(inpt)",
      setup="from __main__ import aug_A",
      globals={
          "image": inpt,
      },
      description='RandomChannelDropout_A',
  ).timeit(1000),
  benchmark.Timer(
      stmt="aug_B(inpt)",
      setup="from __main__ import aug_B",
      globals={
          "image": inpt,
      },
      description='RandomChannelDropout_B',
  ).timeit(1000),
]

compare = benchmark.Compare(results)
compare.print()
```

**Comment by vgilabert94** [comment]
### Benchmark
[Google colab](https://colab.research.google.com/drive/135Ui1jeEleD_M-ks8mlXwlpouE2QdKSw?usp=sharing)

RandomChannelDropout  -> original implementation in this PR.
RandomChannelDropout_v2 -> recommendationsusing only index of channels.

Conclusions:
RandomChannelDropout_v2 is faster in all cases (image_size, dtype, batch, device).

### Results:
This results are RandomChannelDropout and RandomChannelDropout_v2 tested with num_drop_channels=1. 
In colab i have another test with num_drop_channels=3 (black image).

[--------------------------------- cpu ---------------------------------]
                      |  RandomChannelDropout  |  RandomChannelDropout_v2
1 threads: --------------------------------------------------------------
      1-128-float32   |           536.0        |             499.0       
      1-256-float32   |           847.3        |             621.3       
      1-512-float32   |          3124.2        |            1787.6       
      1-1024-float32  |         13190.5        |            4737.5       
      1-2048-float32  |         52327.2        |           31891.5       
      3-128-float32   |           755.2        |             532.0       
      3-256-float32   |          1627.3        |             872.6       
      3-512-float32   |          5164.4        |            2492.2       
      3-1024-float32  |         42339.8        |           23001.7       
      3-2048-float32  |        257906.6        |          113178.5       
      5-128-float32   |          1108.1        |             723.0       
      5-256-float32   |          2662.1        |            1192.0       
      5-512-float32   |         12573.2        |            5687.7       
      5-1024-float32  |         78542.6        |           45207.2       
      5-2048-float32  |        367895.6        |          180389.7       
      1-128-float64   |           565.8        |             478.7       
      1-256-float64   |           937.4        |             640.8       
      1-512-float64   |          2266.9        |            1264.8       
      1-1024-float64  |         12353.2        |            8446.4       
      1-2048-float64  |         85206.5        |           68279.8       
      3-128-float64   |           816.3        |             610.9       
      3-256-float64   |          1808.9        |            1104.3       
      3-512-float64   |          6901.3        |            3347.0       
      3-1024-float64  |         63442.2        |           49897.9       
      3-2048-float64  |        286378.2        |          204398.8       
      5-128-float64   |          1055.4        |             728.0       
      5-256-float64   |          2843.3        |            1660.8       
      5-512-float64   |         14196.1        |            8843.3       
      5-1024-float64  |        105115.4        |           85563.7       
      5-2048-float64  |        482663.8        |          332886.4       
Times are in microseconds (us).

[--------------------------------- cuda --------------------------------]
                      |  RandomChannelDropout  |  RandomChannelDropout_v2
1 threads: --------------------------------------------------------------
      1-128-float32   |           615.8        |            561.2        
      1-256-float32   |           844.0        |            636.7        
      1-512-float32   |          1549.0        |            570.2        
      1-1024-float32  |          4029.2        |            655.7        
      1-2048-float32  |         15197.2        |           1196.6        
      3-128-float32   |           988.5        |            744.9        
      3-256-float32   |          1738.8        |            917.5        
      3-512-float32   |          4525.0        |            917.6        
      3-1024-float32  |         11192.9        |            910.8        
      3-2048-float32  |         74814.2        |           2106.2        
      5-128-float32   |           882.4        |            561.1        
      5-256-float32   |          1851.6        |            578.0        
      5-512-float32   |          5614.6        |            893.5        
      5-1024-float32  |         21454.6        |           1099.3        
      5-2048-float32  |        135010.2        |           3174.3        
      1-128-float64   |           650.6        |            588.4        
      1-256-float64   |           812.3        |            560.5        
      1-512-float64   |          1518.5        |            687.2        
      1-1024-float64  |          4220.3        |            849.3        
      1-2048-float64  |         15507.4        |           1401.6        
      3-128-float64   |           813.4        |            607.5        
      3-256-float64   |          1320.3        |            614.7        
      3-512-float64   |          3313.2        |            741.8        
      3-1024-float64  |         11054.6        |           1228.0        
      3-2048-float64  |         98769.0        |           3205.2        
      5-128-float64   |           915.7        |            601.1        
      5-256-float64   |          1800.7        |            637.7        
      5-512-float64   |          4964.3        |            862.7        
      5-1024-float64  |         23493.8        |           1647.7        
      5-2048-float64  |        138367.8        |           5156.0        
Times are in microseconds (us).

**Comment by edgarriba** [comment]
awesome -- then's let's keep v2

File: kornia/augmentation/_2d/intensity/channel_dropout.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        KORNIA_CHECK(
+            0.0 <= fill_value <= 1.0,
+            f"Invalid value in `fill_value`. Must be a float bewteen 0 and 1. Got: {fill_value}",
+        )
+        self.fill_value = tensor(fill_value)
+
+        KORNIA_CHECK_TYPE(num_drop_channels, int, f"`num_drop_channels` must be an int. Got: {type(num_drop_channels)}")
+        KORNIA_CHECK(
+            1 <= num_drop_channels <= 3,
+            f"Invalid value in `num_drop_channels`. Must be an int bewteen 1 and 3. Got: {num_drop_channels}",
```
Comment (johnnv1) [inline kornia/augmentation/_2d/intensity/channel_dropout.py:78]: why lower than 3? couldn't we support any `C` on this one?

File: tests/augmentation/test_augmentation.py [hunk 3]
```diff
@@ -3509,6 +3510,66 @@ def test_same_on_batch(self, device, dtype):
...
+        with pytest.raises(
+            Exception, match=f"Invalid value in `fill_value`. Must be a float bewteen 0 and 1. Got: {fill_value}"
+        ):
+            RandomChannelDropout(fill_value=fill_value)
+
+        fill_value = 1
+        with pytest.raises(TypeError, match=f"`fill_value` must be a float. Got: {type(fill_value)}"):
+            RandomChannelDropout(fill_value=fill_value)
+
+    @pytest.mark.parametrize("channel_shape, batch_shape", [(3, 1), (3, 2), (3, 5)])
```
Comment (johnnv1) [inline tests/augmentation/test_augmentation.py:3558]:
```suggestion
    @pytest.mark.parametrize("channel_shape, batch_shape", [(3, 1), (1, 1), (3, 5)])
```

File: my_testing.py
Comment (edgarriba) [inline my_testing.py:1]: Don’t include  in the final version

File: my_testing_colormap.py
Comment (edgarriba) [inline my_testing_colormap.py:1]: Or maybe let’s put all this things in the benchmark directory ? @johnnv1

**Review by edgarriba — APPROVED** [review]
You need to git merge origin/main

File: tests/augmentation/test_augmentation.py (hunk 3 above)
Comment (vgilabert94) [reply tests/augmentation/test_augmentation.py:3558]: Why this change?

In local tests are working fine (cpu).

File: kornia/augmentation/_2d/intensity/channel_dropout.py (hunk 2 above)
Comment (vgilabert94) [reply kornia/augmentation/_2d/intensity/channel_dropout.py:78]: I will remove this limitation of 3 channels, and add a KORNIA_CHECK for if num_drop_channels > img.shape[1].

File: my_testing_colormap.py
Comment (vgilabert94) [reply my_testing_colormap.py:1]: I've removed this files in the last commit. Sorry, but are my dirty files to prove the code, nothing interesting to save xD

File: tests/augmentation/test_augmentation.py (hunk 3 above)
Comment (johnnv1) [reply tests/augmentation/test_augmentation.py:3558]: To have tests with other C size

File: my_testing_colormap.py
Comment (johnnv1) [reply my_testing_colormap.py:1]: I think that in the benchmark directory, we should only add things that will be included in the benchmark suite, loose tests/benchmarks should go in the other repo.

@vgilabert94 can you add an benchmark function for this augmentation in [benchmarks/augmentation/2d_intensity_test.py](https://github.com/kornia/kornia/blob/main/benchmarks/augmentation/2d_intensity_test.py)

will looks like
```python
def test_aug_2d_channel_dropout(benchmark, device, dtype, torch_optimizer, shape):
    data = torch.rand(*shape, device=device, dtype=dtype)
    aug = RandomChannelDropout(p=1.0)
    op = torch_optimizer(aug)

    actual = benchmark(op, input=data)

    assert actual.shape == shape
```

File: kornia/augmentation/_2d/intensity/channel_dropout.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    def apply_transform(
+        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any], transform: Optional[Tensor] = None
+    ) -> Tensor:
+        KORNIA_CHECK_SHAPE(input, ["B", "C", "H", "W"])
+        KORNIA_CHECK(
+            self.num_drop_channels <= input.shape[1],
+            "Invalid value in `num_drop_channels`. Cannot be greater than the number of channels of `input`.",
+        )
+
+        out = input.clone()
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/channel_dropout.py:93]: do we really need to clone ?

**Review by edgarriba — APPROVED** [review]
LGTM

**Review by johnnv1 — APPROVED** [review]

File: kornia/augmentation/_2d/intensity/channel_dropout.py (hunk 4 above)
Comment (vgilabert94) [reply kornia/augmentation/_2d/intensity/channel_dropout.py:93]: I think yes for not modifying the original input tensor. The indexing operation returns a view of the original tensor rather than creating a new one.
(If I am wrong, let me know)
